apiserver/pkg/modelservice/modelservice.go

```diff
@@ -0,0 +1,325 @@
+	}
+	return &md
+}
+
+func CreateModelService(ctx context.Context, c dynamic.Interface, input generated.CreateModelServiceInput) (*generated.ModelService, error) {
+	// Overview:
+	// - Get general info from input: displayName, description, types, name & namespace, etc.
+	// - Create *generated.LLM, *generated.embedder accordingly
+	// - Wrap all elements into *generated.ModelService
+	displayName, description, serviceType, APIType := "", "", "", ""
+	var genLLM *generated.Llm
+	var genEmbed *generated.Embedder
+
+	if input.DisplayName != nil {
+		displayName = *input.DisplayName
+	}
+	if input.Description != nil {
+		description = *input.Description
+	}
+	if input.Types != nil {
+		serviceType = *input.Types
+	}
+	if input.APIType != nil {
+		APIType = *input.APIType
+	}
+
+	// FIXME: Codes duplicated with llm & embedder creation.
+	if strings.Contains(serviceType, "llm") {
+		llm := v1alpha1.LLM{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      input.Name,
+				Namespace: input.Namespace,
+			},
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "LLM",
+				APIVersion: v1alpha1.GroupVersion.String(),
+			},
+			Spec: v1alpha1.LLMSpec{
+				CommonSpec: v1alpha1.CommonSpec{
+					DisplayName: displayName,
+					Description: description,
+				},
+				Provider: v1alpha1.Provider{
+					// TODO: Update endpoint info after auth secret creation
+					Enpoint: &v1alpha1.Endpoint{
+						URL: input.Endpoint.URL,
+					},
+				},
+				Type: llms.LLMType(APIType),
+			},
+		}
+
+		// create auth secret
+		if input.Endpoint.Auth != nil {
+			secret := common.MakeAuthSecretName(llm.Name, "llm")
+			err := common.MakeAuthSecret(ctx, c, generated.TypedObjectReferenceInput{
+				Name:      secret,
+				Namespace: &input.Namespace,
+			}, *input.Endpoint.Auth, nil)
+			if err != nil {
+				return nil, err
+			}
+			llm.Spec.Enpoint.AuthSecret = &v1alpha1.TypedObjectReference{
+				Kind:      "Secret",
+				Name:      secret,
+				Namespace: &input.Namespace,
+			}
+		}
+
+		unstructuredLLM, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&llm)
+		if err != nil {
+			return nil, err
+		}
+
+		obj, err := c.Resource(schema.GroupVersionResource{Group: v1alpha1.GroupVersion.Group, Version: v1alpha1.GroupVersion.Version, Resource: "LLM"}).
+			Namespace(input.Namespace).Create(ctx, &unstructured.Unstructured{Object: unstructuredLLM}, metav1.CreateOptions{})
+		if err != nil {
+			return nil, err
+		}
+
+		// update auth secret owner reference
+		if input.Endpoint.Auth != nil {
+			// user obj as the owner
+			secret := common.MakeAuthSecretName(llm.Name, "LLM")
+			err := common.MakeAuthSecret(ctx, c, generated.TypedObjectReferenceInput{
+				Name:      secret,
+				Namespace: &input.Namespace,
+			}, *input.Endpoint.Auth, obj)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		// create *generated.Llm
+		genLLM = obj2LLM(ctx, c, obj)
+	}
+
+	if strings.Contains(serviceType, "embedding") {
+		embedder := v1alpha1.Embedder{
```

Comment:
We can use https://github.com/kubeagi/arcadia/blob/main/apiserver/pkg/embedder/embedder.go#L87 at this place